### DISCUSSION TITLE
How to set node set Ids

### DISCUSSION THREAD

**Discussion by adityaghantasala** [sandialabs/seacas#288] [Q&A] [answered]
Hello I am working on converting some of my input into exodus format. 
Seacas has been very helpful. Thank you for the work. 

Now i am stuck with defining node sets. I have some node sets in my own format. To set these into exodus format, I am using the following code

```python
node_set_ids = range(1,10)
set_nodes = [5,9]
status = my_exo.put_node_set(1, set_nodes)
```

But unfortunately, it results in a an error saying no set with id 1 is found. I tried to set the node set ids but cannot find a suitable function to do. 

Can you point me to the necessary setps to do this ?

Thank you in advance.

**Comment by gsjaardema** [comment, marked as answer]
When you first define the exodus model, you will tell it how many node sets the model will contain (`num_ns`):
```
        >>> status = exo.put_info(title, num_dims, num_nodes, num_elems,
        ...                      num_blocks, num_ns, num_ss)
```
Then, for each of those `num_ns` nodesets, you tell exodus the ID and count of nodes and distribution factors:
```
        >>> exo.put_set_params('EX_NODE_SET', node_set_id,
        ...                 num_ns_nodes, num_ns_dist_facts)
```
You can then use your code above to actually output the `set_nodes`.  Somewhat convulted and not really pythonic, but this is driven by the C API...

Hope that helps.

**Reply by adityaghantasala** [reply]
Hi Thank you for the quick reply. 
Yes This works. 
Thanks again !

**Comment by gsjaardema** [comment]
You might also be interested in using `exomerge` https://gsjaardema.github.io/seacas-docs/exomerge.pdf which is a higher-level interface to the exodus library and provides some higher-level convenience functions.  It has a function 
`create_node_set --- Create a node side from the given list of node indices.` which does most of this in an single step:
```
Example
>>> model.create_node_set(1, [0, 1, 2, 3])
```

**Reply by adityaghantasala** [reply]
I will have a look at this. Thanks for the hint.

**Comment by jessejcarter** [comment]
@gsjaardema Sorry to revive an old thread, but what if the exodus file is already defined as having no nodesets and I want to add one? In my case I'm reading in an exodus file with the exodus python API with no node sets defined and want to add one.

**Reply by gsjaardema** [reply]
Does the method described above in the exomerge.py python interface work for you?

**Reply by jessecarterMOOSE** [reply]
Yes that's what I ended up doing. I think my issue is that once you set the number of some feature in the exodus constructor (numNodes, numNodeSets, etc...), that value can't be modified for the given object. The exomerge API, from my understanding, creates a new exodus object that can hold extra features added after-the-fact.

**Comment by gsjaardema** [comment]
It is very difficult to modify an existing file using the exodus API to add element blocks or sets.  The usual way is to read an existing database, add nodesets/sidesets/blocks and then write a new database.

**Reply by jessecarterMOOSE** [reply]
OK I'll remember that going forward. Thanks
